File: cpp/src/bisection.cpp
#include "bisection.hpp"
#include "calculate_mean.hpp"
#include "sgn.hpp"


double bisect(std::function<double(double)> const& f, double L, double U, double delta)
{
    double const M = calculate_mean(L, U);
    double const vM = f(M);
    if (abs(vM) <= delta) {
        return M;
    }

    if (sgn(f(M)) == sgn(f(L))) {
        L = M;
    } else {
        U = M;
    }

    return bisect(f, L, U, delta);
}

File: python/src/bisection.py
import math

from calculate_mean import calculate_mean


def sign(value):
    return math.copysign(1, value)


def bisect(function, L, U, delta):
    while True:
        M = calculate_mean(L, U)
        vM = function(M)
        if abs(vM) < delta:
            return M

        if sign(vM) == sign(function(L)):
            L = M
        else:
            U = M
